File: android/src/main/kotlin/com/sabiantech/sabian_native_common/channels/events/EventChannelHandler.kt
package com.sabiantech.sabian_native_common.channels.events

import android.util.Log
import com.sabiantech.sabian_native_common.channels.ChannelPayload
import com.sabiantech.sabian_native_common.channels.IChannelHandler
import com.sabiantech.sabian_native_common.events.ErrorEvent
import com.sabiantech.sabian_native_common.events.LoaderEvent
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.EventChannel.StreamHandler

class EventChannelHandler : IChannelHandler<StreamHandler, EventChannelPayload>() {
    override fun register() {
        register("com.sabian_common_native.events.loader", LoaderEvent)
        register("com.sabian_common_native.events.error", ErrorEvent)
    }

    override fun create(key: String, payload: ChannelPayload) {
        super.create(key, payload)
        channelHandlers.forEach { (t, u) ->
            execute(u, EventChannelPayload(t, payload.messenger))
        }
    }

    override fun execute(action: StreamHandler, args: EventChannelPayload) {
        Log.e("sabian_common_native", "Executing event for ${args.channelName}")
        EventChannel(args.messenger, args.channelName).setStreamHandler(
                action
        )
    }
}

File: ios/Classes/channels/events/EventChannelHandler.swift
//
//  EventChannelHandler.swift
//  Runner
//
//  Created by bryosabian on 22/11/2023.
//

import Foundation
import Flutter

class EventChannelHandler : IChannelHandler<AbstractEvent, EventChannelPayload> {
    
    override func register() {
        register("com.sabian_common_native.events.loader", LoaderEvent.shared)
        register("com.sabian_common_native.events.error", ErrorEvent.shared)
    }
    
    override func create(_ key: String, payload : ChannelPayload) {
        super.create(key, payload: payload)
        self.map.forEach({ [unowned self] key,value in
            self.execute(value, EventChannelPayload(channelName: key,messenger: payload.messengaer))
        })
    }
    
    

    override func execute(_ action: AbstractEvent, _ args: EventChannelPayload) {
        print("sabian_native : Executing event for \(args.channelName)")
        FlutterEventChannel(name: args.channelName, binaryMessenger: args.messenger).setStreamHandler(action)
    }
}
